src/pip/_internal/index.py

```diff
@@ -386,12 +387,34 @@ def iter_applicable(self):
         # Again, converting version to str to deal with debundling.
         return (c for c in self.iter_all() if str(c.version) in self._versions)
 
-    def get_best(self):
-        # type: () -> Optional[InstallationCandidate]
+    def get_best(self, hashes=None):
+        # type: (Optional[Hashes]) -> Optional[InstallationCandidate]
         """Return the best candidate available, or None if no applicable
         candidates are found.
         """
         candidates = list(self.iter_applicable())
+        if hashes:
+            # If we are in hash-checking mode, filter out candidates that will
+            # fail the hash check per the hash provided in their Link URL.
+            # This prevents HashMismatch errors when a new distribution is
+            # uploaded for an old release.
+            # This is not a security check: after download the contents will
+            # be hashed and compared to the known-good hashes.
+            def test_against_hashes(candidate):
+                link = candidate.location
+                if not link.hash:
+                    # We can only filter candidates with hashes in their URLs.
+                    return True
+                is_match = hashes.test_against_hash(link.hash_name, link.hash)
+                if not is_match:
+                    logger.warning(
+                        "candidate %s ignored: hash %s:%s not among provided "
+                        "hashes",
+                        link.filename, link.hash_name, link.hash,
+                    )
+                return is_match
+
+            candidates = [c for c in candidates if test_against_hashes(c)]
```

Comment:
This combined with the `list()` call above unnecessarily loops through the candidate list twice.

Reply:
I think instead of filtering, I want to only test the best value (in the common case where the best value's hash is correct). This would be both more performant and avoid spammy warnings.